### PR TITLE
Persistence Component: Bulk Load Improvements + Metadata Table Changes

### PR DESCRIPTION
**Pull request by kumuwu**
#### What type of PR is this?

- Improvement
- Bug Fix

#### What does this PR do / why is it needed ?

1. Unify metadata table such that all ingest modes use the same metadata table
2. Change the unified metadata table's schema, with a batch_source_info JSON column as well as a additional_metadata JSON column
3. Implement schema evolution for metadata table
4. Implement fields to exclude from digest for UDF based digest generation
5. Change Big Query's Bulk Load to use external table
6. Enrich Snowflake's Bulk Load's error message to return more information
7. Improve H2's Bulk Load stats to use batch ID instead of batch timestamp
8. Allow users to configure the string value of batch success status
9. Bug fix: add missing delete indicator conditions for nontemporal delta

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java [hunk 1]
```diff
@@ -75,15 +69,13 @@ class BulkLoadPlanner extends Planner
...
-        bulkLoadEventIdValue = plannerOptions.bulkLoadEventIdValue();
         stagedFilesDataset = (StagedFilesDataset) datasets.stagingDataset();
-        bulkLoadMetadataDataset = bulkLoadMetadataDataset().orElseThrow(IllegalStateException::new);
 
         transformWhileCopy = capabilities.contains(Capability.TRANSFORM_WHILE_COPY);
         if (!transformWhileCopy)
         {
-            tempDataset = DatasetDefinition.builder()
-                .schema(datasets.stagingDataset().schema())
+            tempDataset = ExternalDataset.builder()
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java:77]: Let's rename this to externalDataset

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java [hunk 2]
```diff
@@ -225,25 +204,34 @@ List<String> getDigestOrRemainingColumns()
...
     public LogicalPlan buildLogicalPlanForMetadataIngest(Resources resources)
     {
-        BulkLoadMetadataUtils bulkLoadMetadataUtils = new BulkLoadMetadataUtils(bulkLoadMetadataDataset);
-        String batchSourceInfo = jsonifyBatchSourceInfo(stagedFilesDataset.stagedFilesDatasetProperties());
-        StringValue datasetName = StringValue.of(mainDataset().datasetReference().name());
-        Insert insertMetaData = bulkLoadMetadataUtils.insertMetaData(datasetName, StringValue.of(batchSourceInfo));
-        return LogicalPlan.of(Arrays.asList(insertMetaData));
+        // Save file paths/patterns and event id into batch_source_info column
+        Map<String, Object> batchSourceInfoMap = new HashMap<>();
+        LogicalPlanUtils.jsonifyBulkLoadSourceInfo(batchSourceInfoMap, stagedFilesDataset.stagedFilesDatasetProperties(), options().bulkLoadEventIdValue());
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java:209]: We can let the jsonifyBulkLoadSourceInfo return the json as a map

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java [hunk 3]
```diff
@@ -257,8 +252,30 @@ protected PlannerOptions options()
...
     public LogicalPlan buildLogicalPlanForMetadataIngest(Resources resources)
     {
-        List<DatasetFilter> stagingFilters = LogicalPlanUtils.getDatasetFilters(originalStagingDataset());
-        return LogicalPlan.of(Arrays.asList(metadataUtils.insertMetaData(mainTableName, batchStartTimestamp, batchEndTimestamp, stagingFilters)));
+        // Save staging filters into batch_source_info column
+        Map<String, Object> batchSourceInfoMap = new HashMap<>();
+        if (originalStagingDataset() instanceof DerivedDataset)
+        {
+            DerivedDataset derivedDataset = (DerivedDataset) originalStagingDataset();
+            LogicalPlanUtils.jsonifyStagingFilters(batchSourceInfoMap, derivedDataset.datasetFilters());
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java:260]: We can let the jsonifyStagingFilters return the json as a map

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java [hunk 4]
```diff
@@ -257,8 +252,30 @@ protected PlannerOptions options()
...
+
+        // Save additional metadata into additional_metadata column
+        Optional<StringValue> additionalMetadata = Optional.empty();
+        if (!options().additionalMetadata().isEmpty())
+        {
+            additionalMetadata = Optional.of(LogicalPlanUtils.getStringValueFromMap(options().additionalMetadata()));
+        }
+
+        // Save "DONE" into status column
+        StringValue status = StringValue.of(MetadataUtils.MetaTableStatus.DONE.toString()); // todo: may be a good chance to unify the status now
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java:276]: Let's get the config from user - batchSuccessStatusValue.
We can keep the default as DONE.

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-snowflake/src/main/java/org/finos/legend/engine/persistence/components/relational/snowflake/SnowflakeSink.java [hunk 5]
```diff
@@ -269,15 +283,15 @@ public IngestorResult performBulkLoad(Datasets datasets, Executor<SqlGen, Tabula
...
+        if (dataFilePathsWithErrors.isEmpty())
         {
             result = resultBuilder
                 .status(IngestStatus.SUCCEEDED)
                 .build();
         }
         else
         {
-            String errorMessage = String.format("Unable to bulk load these files: %s", String.join(",", dataFilePathsWithFailedBulkLoad));
+            String errorMessage = String.join(",", errorMessages);
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-snowflake/src/main/java/org/finos/legend/engine/persistence/components/relational/snowflake/SnowflakeSink.java:294]: Let's log this error

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java (hunk 1 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java]: Sure, changed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java (hunk 2 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java]: Sure, changed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java (hunk 3 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java]: Sure, changed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java (hunk 4 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java]: Sure, changed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-snowflake/src/main/java/org/finos/legend/engine/persistence/components/relational/snowflake/SnowflakeSink.java (hunk 5 above)
Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-snowflake/src/main/java/org/finos/legend/engine/persistence/components/relational/snowflake/SnowflakeSink.java]: Sure, changed

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java [hunk 6]
```diff
@@ -225,25 +204,25 @@ List<String> getDigestOrRemainingColumns()
...
     public LogicalPlan buildLogicalPlanForMetadataIngest(Resources resources)
     {
-        BulkLoadMetadataUtils bulkLoadMetadataUtils = new BulkLoadMetadataUtils(bulkLoadMetadataDataset);
-        String batchSourceInfo = jsonifyBatchSourceInfo(stagedFilesDataset.stagedFilesDatasetProperties());
-        StringValue datasetName = StringValue.of(mainDataset().datasetReference().name());
-        Insert insertMetaData = bulkLoadMetadataUtils.insertMetaData(datasetName, StringValue.of(batchSourceInfo));
-        return LogicalPlan.of(Arrays.asList(insertMetaData));
+        // Save file paths/patterns and event id into batch_source_info column
+        Map<String, Object> batchSourceInfoMap = LogicalPlanUtils.jsonifyBulkLoadSourceInfo(stagedFilesDataset.stagedFilesDatasetProperties(), options().bulkLoadEventIdValue());
+        Optional<StringValue> batchSourceInfo = LogicalPlanUtils.getStringValueFromMapIfNotEmpty(batchSourceInfoMap);
```
Comment (prasar-ashutosh) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java:209]: Let's rename this to : getStringValueFromMap

Comment (kumuwu) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/BulkLoadPlanner.java]: Sure, changed!

**Pull request merged**